File: i3rd/UI/PropertyListBasedViewController.m

#import "PropertyListBasedViewController.h"

NSString *const PropertyListBasedViewControllerSectionsKey = @"Sections";
NSString *const PropertyListBasedViewControllerSectionTitleKey = @"SectionTitle";
NSString *const PropertyListBasedViewControllerRowsKey = @"Rows";
NSString *const PropertyListBasedViewControllerRowImageKey = @"RowImage";
NSString *const PropertyListBasedViewControllerRowTitleKey = @"RowTitle";
NSString *const PropertyListBasedViewControllerRowDetailKey = @"RowDetail";
NSString *const PropertyListBasedViewControllerNextKey = @"Next";
NSString *const PropertyListBasedViewControllerPresentedKey = @"Presented";

@implementation PropertyListBasedViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.tableView.estimatedRowHeight = 44;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    [self.tableView registerNib:[UINib nibWithNibName:@"PropertyListBasedTableViewCell" bundle:nil] forCellReuseIdentifier:@"PropertyListBasedTableViewCell"];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (void)configureCell:(PropertyListBasedTableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSDictionary *sectionInfo = [self.sections objectAtIndex:indexPath.section];
    NSArray *rows = [sectionInfo objectForKey:PropertyListBasedViewControllerRowsKey];
    NSDictionary *rowInfo = [rows objectAtIndex:indexPath.row];
    NSString *rowImage = [rowInfo objectForKey:PropertyListBasedViewControllerRowImageKey];
    NSString *rowTitle = [rowInfo objectForKey:PropertyListBasedViewControllerRowTitleKey];
    NSString *rowDetail = [rowInfo objectForKey:PropertyListBasedViewControllerRowDetailKey];
    id next = [rowInfo objectForKey:PropertyListBasedViewControllerNextKey];
    id presented = [rowInfo objectForKey:PropertyListBasedViewControllerPresentedKey];
    
    if (next) {
        
    } else if (presented) {
        
    } else {
        
    }
    
    cell.leftImageView.image = [UIImage imageNamed:rowImage];
    cell.leftLabel.text = rowTitle;
    cell.rightLabel.text = rowDetail;
}

#pragma mark - Table View Data Source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [self.sections count];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    NSDictionary *sectionInfo = [self.sections objectAtIndex:section];
    NSString *sectionTitle = [sectionInfo objectForKey:PropertyListBasedViewControllerSectionTitleKey];
    return sectionTitle;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSDictionary *sectionInfo = [self.sections objectAtIndex:section];
    NSArray *rows = [sectionInfo objectForKey:PropertyListBasedViewControllerRowsKey];
    return [rows count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    PropertyListBasedTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"PropertyListBasedTableViewCell" forIndexPath:indexPath];
    [self configureCell:cell forRowAtIndexPath:indexPath];
    return cell;
}

#pragma mark - Table View Delegate

//- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
//{
//    static PropertyListBasedTableViewCell *cell = nil;
//    static dispatch_once_t onceToken;
//    dispatch_once(&onceToken, ^{
//        cell = [[[UINib nibWithNibName:@"Cell" bundle:nil] instantiateWithOwner:nil options:nil] firstObject];
//    });
//    
//    CGRect cellFrame = cell.frame;
//    cellFrame.size.width = tableView.frame.size.width;
//    cell.frame = cell.frame;
//    
//    [self configureCell:cell forRowAtIndexPath:indexPath];
//    [cell setNeedsLayout];
//    [cell layoutIfNeeded];
//    
//    CGSize size = [cell.contentView systemLayoutSizeFittingSize:UILayoutFittingCompressedSize];
//    return size.height + 1.0;
//}

- (NSIndexPath *)tableView:(UITableView *)tableView willSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    PropertyListBasedTableViewCell *cell = (PropertyListBasedTableViewCell *)[tableView cellForRowAtIndexPath:indexPath];
    if ([cell canBeSelected]) {
        return indexPath;
    }
    return nil;
}

#pragma mark - Data Source Model Association

- (NSString *)modelIdentifierForElementAtIndexPath:(NSIndexPath *)indexPath inView:(UIView *)view
{
    return nil;
}

- (NSIndexPath *)indexPathForElementWithModelIdentifier:(NSString *)identifier inView:(UIView *)view
{
    NSArray *components = [identifier componentsSeparatedByString:@", "];
    NSInteger section = [components[0] integerValue];
    NSInteger row = [components[1] integerValue];
    return [NSIndexPath indexPathForRow:section inSection:row];
}

@end

File: MaChérie/UI/PropertyListBasedViewController.swift
//
//  PropertyListBasedViewController.swift
//  MaChérie
//
//  Created by Leon Li on 2018/6/14.
//  Copyright © 2018 Leon & Vane. All rights reserved.
//

import UIKit

let PropertyListBasedViewControllerSectionsKey = "Sections"
let PropertyListBasedViewControllerSectionTitleKey = "SectionTitle"
let PropertyListBasedViewControllerRowsKey = "Rows"
let PropertyListBasedViewControllerRowImageKey = "RowImage"
let PropertyListBasedViewControllerRowTitleKey = "RowTitle"
let PropertyListBasedViewControllerRowDetailKey = "RowDetail"
let PropertyListBasedViewControllerNextKey = "Next"
let PropertyListBasedViewControllerPresentedKey = "Presented"

class PropertyListBasedViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, UIDataSourceModelAssociation {
    @IBOutlet var tableView: UITableView!
    
    @objc var sections = NSArray()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.estimatedRowHeight = 44
        tableView.register(UINib(nibName: "PropertyListBasedTableViewCell", bundle: nil), forCellReuseIdentifier: "PropertyListBasedTableViewCell")
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if let indexPath = tableView.indexPathForSelectedRow {
            tableView.deselectRow(at: indexPath, animated: true)
        }
    }
    
    func configureCell(_ cell: PropertyListBasedTableViewCell, forRowAt indexPath: IndexPath) {
        let sectionInfo = sections[indexPath.section] as? NSDictionary
        let rows = sectionInfo?[PropertyListBasedViewControllerRowsKey] as? NSArray
        let rowInfo = rows?[indexPath.row] as? NSDictionary
        let rowImage = rowInfo?[PropertyListBasedViewControllerRowImageKey] as? String
        let rowTitle = rowInfo?[PropertyListBasedViewControllerRowTitleKey] as? String
        let rowDetail = rowInfo?[PropertyListBasedViewControllerRowDetailKey] as? String
        
        cell.leftImageView.image = rowImage != nil ? UIImage(named: rowImage!) : nil
        cell.leftLabel.text = rowTitle
        cell.rightLabel.text = rowDetail
    }
    
    // MARK: - Table View Data Source
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return sections.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionInfo = sections[section] as! NSDictionary
        let rows = sectionInfo[PropertyListBasedViewControllerRowsKey] as! NSArray
        return rows.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "PropertyListBasedTableViewCell", for: indexPath) as! PropertyListBasedTableViewCell
        configureCell(cell, forRowAt: indexPath)
        return cell
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let sectionInfo = sections[section] as! NSDictionary
        let sectionTitle = sectionInfo[PropertyListBasedViewControllerSectionTitleKey] as? String
        return sectionTitle
    }
    
    // MARK: - Table View Delegate
    
    func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
        let cell = tableView.cellForRow(at: indexPath) as! PropertyListBasedTableViewCell
        return cell.canBeSelected() ? indexPath : nil
    }
    
    // MARK: - Data Source Model Association
    
    func modelIdentifierForElement(at idx: IndexPath, in view: UIView) -> String? {
        return nil
    }
    
    func indexPathForElement(withModelIdentifier identifier: String, in view: UIView) -> IndexPath? {
        let components = identifier.components(separatedBy: ", ")
        let section = Int(components[0])!
        let row = Int(components[1])!
        return IndexPath(row: row, section: section)
    }
}
